docs/docs/routing.md

```diff
@@ -0,0 +1,261 @@
+
+Secondly, navigation events are only emitted when they're caused by external stimuli such as the *Back* or *Forward* buttons. When the application itself updates the URL, nothing will happen. This means that the application will need to manually pass the new URL from the link that caused it to the top-level component that is handling routing.
+
+Typically, these issues are solved by creating a separate object which manages access to the HTML5 History API. This object will do two things:
+
+- It will normalise events between browsers
+- It will emit navigation events caused by both the application and external stimuli
+
+While it is certainly possible to create your own object to manage access to the History API, in this guide we'll use the popular [history](https://github.com/mjackson/history) package.
+
+```bash
+npm install --save history
+```
+
+As the History API is global, our object will also need to be accessible everywhere. For example, we can create a separate file and export the object from it, so that any other modules can later import it.
+
+```js
+import createBrowserHistory from 'history/createBrowserHistory';
+
+const browserHistory = createBrowserHistory();
+
+export default browserHistory;
+```
+
+This guide will use the following methods and properties from the `browserHistory` object:
+
+-   `browserHistory.location`
+
+    An object that contains the browser's current location. Its shape is similar to the object available at `window.location`:
+
+    ```js
+    {
+      pathname: '/friends',
+      query: '?page=1',
+      hash: '#top'
+    }
+    ```
+
+-   `browserHistory.listen(callbackFunction)`
+    
+    Subscribes to history events, returning a function that will cancel the subscription when called.
+
+    While subscribed, the callback function will be called with the latest `location` each time the user navigates. For example:
+
+    ```js
+    const unlisten = browserHistory.listen((location) => {
+      console.log(location.pathname);
+    });
+    
+-   `browserHistory.push(location)`
+
+    Adds a new entry to the browser history using `window.history.pushState()`, then calls any listeners registered with `browserHistory.listen()`
+
+You can view the documentation for the full `browserHistory` object API at the [history](https://github.com/mjackson/history) package's page.
+
+## Links
+
+By default, HTML `<a>` elements will always cause the page to reload when clicked. If you'd like to update the URL using the History API, you'll need to cancel the default behaviour and then implement an alternative.
+
+You could do this by adding an `onClick` handler to each individual `<a>` element, but since most links within an app will share the same logic, people often create a component for them. Instances of this component behave exactly like `<a>` elements, except that instead of refreshing the page, they will update the URL in-place using `browserHistory.push()`:
+
+```js{47}
+class Link extends React.Component {
```

Comment:
@mjackson I learned a lot from your code and really grateful that you've made it available. I definitely did not intend to copy/paste without attribution.

The `Link` component does contain a lot of similar logic in the `onChange` handler, as it performs the same task -- but it is not a direct copy/paste. This is what I meant by "based on". But if you feel like it is a copyright violation, I will remove this PR and write a new `onChange` handler in my own library. Please let me know.